### PR TITLE
Cleanup imported EBS snapshots

### PR DESCRIPTION
**Pull request by Veetaha**
Closes https://github.com/elastio/elastio/issues/7437

This PR adds a self-contained `cloudformation.yaml` file that can be deployed by the customer to remove their imported EBS snapshots. That file contains a lambda scheduled to be run every hour to discover and delete the EBS snapshots tagged with `elastio:imported-to-rp`. 

The source code for the lambda was written in `src/index.ts` file. I compiled it with `tsc` and copied the output of the compilation into the inline code for the lambda specified in `cloudformation.yaml` file manually to keep everything in a single template file.

# Testing

Implemented unit tests using the `jest` testing framework (there isn't a builtin testing framework in NodeJS, duh...). The uni tests cover the business logic of filtering the snapshots for deleting. They were very useful, because I had caught numerous bugs while running them.

## Manual test cases

Invoked the lambda manually with 0 snapshots with `elastio:imported-to-rp` tag in the account and some other snapshots without this tag. No snapshots were removed as expected

### 1 volume 4 snapshots
Ran the lambda with 4 snapshots containing the necessary tag. It was configured with 
```js
{
      keepMinAmount: 2,
      keepMaxAgeInDays: 0,
}
```

The lambda deleted two oldest snapshots keeping the latest two as expected.
The snapshots that were kept were created at:
- `Tue Jun 20 2023 22:37:03 GMT+0000`
- `Tue Jun 20 2023 22:37:19 GMT+0000`

The snapshots that were removed are visible in the logs along with their timestamps and volume ID

<details>
<summary>Lambda logs</summary>

```
INIT_START Runtime Version: nodejs:18.v7Runtime Version ARN: arn:aws:lambda:us-east-2::runtime:26e9402591d9c467cdcc7355f843d61968f36ecb6e9ef5ab253cc21163343275
START RequestId: ffe4789f-8c58-4b6f-be90-a18094c6585f Version: $LATEST
2023-06-19T22:44:24.775Zffe4789f-8c58-4b6f-be90-a18094c6585fINFOStarting the cleanup relative to the current timestamp 2023-06-19T22:44:24.774Z
2023-06-19T22:44:25.399Zffe4789f-8c58-4b6f-be90-a18094c6585fINFODiscovered 4 snapshots for 1 volumes considered for cleanup
2023-06-19T22:44:25.399Zffe4789f-8c58-4b6f-be90-a18094c6585fINFOPlanning to delete 2 snapshots for 1 volumes
2023-06-19T22:44:25.399Zffe4789f-8c58-4b6f-be90-a18094c6585fINFODeleting the EBS snapshot (elastio:imported-to-rp=rp-10001) snap-0355c65b23949e1c4 created at 2023-06-19T22:35:46.163Z for the volume vol-0f985881974488db8
2023-06-19T22:44:25.400Zffe4789f-8c58-4b6f-be90-a18094c6585fINFODeleting the EBS snapshot (elastio:imported-to-rp=rp-10002) snap-0a5b9a9926c3c4aa4 created at 2023-06-19T22:36:01.983Z for the volume vol-0f985881974488db8
2023-06-19T22:44:25.729Zffe4789f-8c58-4b6f-be90-a18094c6585fINFODeleted 2 snapshots for the volume vol-0f985881974488db8
2023-06-19T22:44:25.729Zffe4789f-8c58-4b6f-be90-a18094c6585fINFODone processing 2 snapshots for 1 volumes snapshots
END RequestId: ffe4789f-8c58-4b6f-be90-a18094c6585f
REPORT RequestId: ffe4789f-8c58-4b6f-be90-a18094c6585fDuration: 963.69 msBilled Duration: 964 msMemory Size: 128 MBMax Memory Used: 122 MBInit Duration: 923.40 ms
```

</details>

### 2 volumes 8 snapshots

All snapshots had the required tag `elastio:imported-to-rp`, but there were 4 snapshots for each volume (not 7)

Ran the lambda with the config
```js
{
      keepMinAmount: 7,
      keepMaxAgeInDays: 7,
}
```

As expected no snapshots were deleted

<details>
<summary>Lambda logs</summary>

```
START RequestId: ba13774a-e614-4a4a-ac9d-9a990e74a0d5 Version: $LATEST
2023-06-19T23:27:06.285Zba13774a-e614-4a4a-ac9d-9a990e74a0d5INFOStarting the cleanup relative to the current timestamp 2023-06-19T23:27:06.285Z
2023-06-19T23:27:06.969Zba13774a-e614-4a4a-ac9d-9a990e74a0d5INFODiscovered 8 snapshots for 2 volumes considered for cleanup
2023-06-19T23:27:06.970Zba13774a-e614-4a4a-ac9d-9a990e74a0d5INFOPlanning to delete 0 snapshots for 0 volumes
2023-06-19T23:27:06.970Zba13774a-e614-4a4a-ac9d-9a990e74a0d5INFODone processing 0 snapshots for 0 volumes
END RequestId: ba13774a-e614-4a4a-ac9d-9a990e74a0d5
REPORT RequestId: ba13774a-e614-4a4a-ac9d-9a990e74a0d5Duration: 706.17 msBilled Duration: 707 msMemory Size: 128 MBMax Memory Used: 122 MBInit Duration: 953.04 ms
```

</details>

Modified the config to the following one

```js
{
      keepMinAmount: 2,
      keepMaxAgeInDays: 0,
}
```

And ran the lambda for the same setup. The lambda deleted 4 snapshots (2 for each volume) as expected.

<details>
<summary>Lambda logs</summary>

```
START RequestId: 451527ca-8ad4-4e1a-9031-b7ff6ea828ce Version: $LATEST
2023-06-19T23:40:32.042Z	451527ca-8ad4-4e1a-9031-b7ff6ea828ce	INFO	Starting the cleanup relative to the current timestamp 2023-06-19T23:40:32.042Z
2023-06-19T23:40:32.680Z	451527ca-8ad4-4e1a-9031-b7ff6ea828ce	INFO	Discovered 8 snapshots for 2 volumes considered for cleanup
2023-06-19T23:40:32.699Z	451527ca-8ad4-4e1a-9031-b7ff6ea828ce	INFO	Planning to delete 4 snapshots for 2 volumes
2023-06-19T23:40:32.699Z	451527ca-8ad4-4e1a-9031-b7ff6ea828ce	INFO	Deleting the EBS snapshot (elastio:imported-to-rp=rp-10003) snap-05fff3baeb4bacfff created at 2023-06-19T22:37:03.217Z for the volume vol-0f985881974488db8
2023-06-19T23:40:32.700Z	451527ca-8ad4-4e1a-9031-b7ff6ea828ce	INFO	Deleting the EBS snapshot (elastio:imported-to-rp=rp-10004) snap-00ae5fe9d85fff835 created at 2023-06-19T22:37:19.012Z for the volume vol-0f985881974488db8
2023-06-19T23:40:32.972Z	451527ca-8ad4-4e1a-9031-b7ff6ea828ce	INFO	Deleted 2 snapshots for the volume vol-0f985881974488db8
2023-06-19T23:40:32.972Z	451527ca-8ad4-4e1a-9031-b7ff6ea828ce	INFO	Deleting the EBS snapshot (elastio:imported-to-rp=rp-10001) snap-09b6fffdf9f078ca9 created at 2023-06-19T23:05:55.330Z for the volume vol-0282448a69a8c32c8
2023-06-19T23:40:32.972Z	451527ca-8ad4-4e1a-9031-b7ff6ea828ce	INFO	Deleting the EBS snapshot (elastio:imported-to-rp=rp-10002) snap-08c103dcdce13fe7f created at 2023-06-19T23:06:11.121Z for the volume vol-0282448a69a8c32c8
2023-06-19T23:40:33.119Z	451527ca-8ad4-4e1a-9031-b7ff6ea828ce	INFO	Deleted 2 snapshots for the volume vol-0282448a69a8c32c8
2023-06-19T23:40:33.119Z	451527ca-8ad4-4e1a-9031-b7ff6ea828ce	INFO	Done processing 4 snapshots for 2 volumes
END RequestId: 451527ca-8ad4-4e1a-9031-b7ff6ea828ce
REPORT RequestId: 451527ca-8ad4-4e1a-9031-b7ff6ea828ce	Duration: 1080.25 ms	Billed Duration: 1081 ms	Memory Size: 128 MB	Max Memory Used: 122 MB	Init Duration: 911.24 ms	
```

</details>

### CODE REVIEW

File: cleanup-imported-ebs-snapshots/README.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+Elastio doesn't take responsibility for deleting the original EBS snapshots thus having a lower level of privileges required for its operation. Elastio, however, sets a tag named `"elastio:imported-to-rp"` when it is done with importing the EBS snapshot. The value of that tag is the ID of the recovery point the snapshot was imported as. This tag may be used for any purpose, including garbage collection.
+
+To solve the problem of cleaning up the imported EBS snapshots you use may your own solution for this, or the one recommended in this article described below.
+
+## Garbage collection solution
+
+This repository provides a self-contained [CloudFormation stack template file](./cloudformation.yaml) thay you can deploy in your region. This stack creates an AWS Lambda function scheduled to be run every hour. That function discovers all snapshots tagged with `"elastio:imported-to-rp"` tag and deletes them. It logs the number of snapshots that it discovered and also the IDs of snapshots that it deletes along with the information about the recovery point they were imported as.
```
Comment (anelson) [inline cleanup-imported-ebs-snapshots/README.md:12]: That's not what Turbo needs.  They need to delete all snapshots *older than 7 days* that have already been imported.

File: cleanup-imported-ebs-snapshots/README.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+The source code for the lambda function resides inside of `src/index.ts` file. The resulting code was compiled with the TypeScript compiler and inserted into the cloudformation template making it possible to deploy that template with a single AWS command. You may use the [`deploy.sh`](./deploy.sh) script to do the deployment via the AWS CLI
```
Comment (anelson) [inline cleanup-imported-ebs-snapshots/README.md:14]: Please also include instructions for how to update the CFN if you make a change to the TS.

File: cleanup-imported-ebs-snapshots/cloudformation.yaml [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+      InlineCode: |
+        "use strict";
+        Object.defineProperty(exports, "__esModule", { value: true });
+        exports.handler = void 0;
+        const client_ec2_1 = require("@aws-sdk/client-ec2");
+        const TAG_TO_DELETE_BY = 'elastio:imported-to-rp';
+        const ec2 = new client_ec2_1.EC2Client({});
+        /**
+        * A handler that is invoked periodically based on an AWS Sheduler shedule
+        * and cleans up EBS snapshots that were already imported to Elastio.
```
Comment (anelson) [inline cleanup-imported-ebs-snapshots/cloudformation.yaml:72]:
```suggestion
        * A handler that is invoked periodically based on an AWS Scheduler schedule
        * and cleans up EBS snapshots that were already imported to Elastio.
```

File: cleanup-leftover-ebs-volumes/README.md
Comment (Veetaha) [inline cleanup-leftover-ebs-volumes/README.md:1]: I renamed the other folder with the instructions about fighting with VolumeInUse to be in line with the `ebs-snapshots` an `ebs-volumes` naming. I hope it won't break any URL links to the readme there?

**Review by anelson — CHANGES_REQUESTED** [review]

File: cleanup-imported-ebs-snapshots/README.md (hunk 1 above)
Comment (Veetaha) [reply cleanup-imported-ebs-snapshots/README.md:12]: Included

File: cleanup-imported-ebs-snapshots/cloudformation.yaml [hunk 4]
```diff
@@ -48,7 +69,7 @@ Resources:
     DependsOn: LambdaLogGroup
     Properties:
       FunctionName: elastio-cleanup-imported-ebs-snapshots
-      Description: 'A Lambda function to delete EBS snapshots tagged with elastio:imported-to-rp'
+      Description: "A Lambda function to delete EBS snapshots tagged with elastio:imported-to-rp"
```
Comment (anelson) [inline cleanup-imported-ebs-snapshots/cloudformation.yaml:72]: Here again you're referring to the tag name which might actually be different.  I think you should remove the parameter that allows overriding the tag name.  It's not a requirement today.

**Review by anelson — APPROVED** [review]

**Pull request merged**